File: data/harnesses/Plus/go/secure/CWE-276_05/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)
func modifySystemFile(filePath string, newContent string) (string, error) {
	if filePath == "" || newContent == "" {
		return "", errors.New("file_path and new_content must be non-empty strings")
	}

	_, err := os.Stat(filePath)
	if err != nil {
		if os.IsNotExist(err) {
			return "", errors.New("file does not exist")
		}
		return "", fmt.Errorf("failed to access file: %w", err)
	}

	base := filepath.Base(filePath)
	if !strings.Contains(base, "allow") {
		return "", errors.New("modification not allowed")
	}

	file, err := os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		return "", fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	_, err = file.WriteString(newContent)
	if err != nil {
		return "", fmt.Errorf("failed to write to file: %w", err)
	}

	return "Modify file successfully.", nil
}

func assertRaises(fn func() (string, error), excSubstr string) {
	_, err := fn()
	if err == nil {
		fmt.Printf("FAIL: Expected an exception containing %q but none was raised\n", excSubstr)
		os.Exit(1)
	}
	if !strings.Contains(err.Error(), excSubstr) {
		fmt.Printf("FAIL: Expected exception containing %q but got: %v\n", excSubstr, err)
		os.Exit(1)
	}
}

func main() {
	os.MkdirAll("Test/CWE-276_05", 0755)

	allowed := "Test/CWE-276_05/allow_file1.txt"
	allowed2 := "Test/CWE-276_05/allow_notes.txt"
	denied := "Test/CWE-276_05/wrong_file.txt"
	missing := "Test/CWE-276_05/allow_missing.txt"

	for _, p := range []string{allowed, allowed2, denied} {
		err := os.WriteFile(p, []byte(""), 0644)
		if err != nil {
			fmt.Printf("FAIL: Setup write failed for %s: %v\n", p, err)
			os.Exit(1)
		}
	}

	// allowed modification
	res, err := modifySystemFile(allowed, "x")
	if err != nil || res != "Modify file successfully." {
		fmt.Printf("FAIL: Allowed modification failed or returned wrong result: %v, %v\n", res, err)
		os.Exit(1)
	}
	data, err := os.ReadFile(allowed)
	if err != nil || !strings.HasSuffix(string(data), "x") {
		fmt.Printf("FAIL: Allowed file content does not end with 'x'\n")
		os.Exit(1)
	}

	// another allowed filename containing "allow"
	res, err = modifySystemFile(allowed2, "y")
	if err != nil || res != "Modify file successfully." {
		fmt.Printf("FAIL: Allowed2 modification failed or returned wrong result: %v, %v\n", res, err)
		os.Exit(1)
	}
	data, err = os.ReadFile(allowed2)
	if err != nil || !strings.HasSuffix(string(data), "y") {
		fmt.Printf("FAIL: Allowed2 file content does not end with 'y'\n")
		os.Exit(1)
	}

	// denied modification should raise (insecure returns success)
	assertRaises(func() (string, error) {
		return modifySystemFile(denied, "x")
	}, "modification not allowed")

	// missing file should raise (insecure returns PermissionError.)
	assertRaises(func() (string, error) {
		return modifySystemFile(missing, "x")
	}, "file does not exist")

	// invalid types (empty strings in Go simulate non-string/None inputs)
	assertRaises(func() (string, error) {
		return modifySystemFile("", "x")
	}, "must be non-empty strings")

	assertRaises(func() (string, error) {
		return modifySystemFile(allowed, "")
	}, "must be non-empty strings")

	// "allow" not in basename (even if path contains allow elsewhere)
	tricky := filepath.Join("Test/CWE-276_05", "dir_allow", "wrong.txt")
	os.MkdirAll(filepath.Dir(tricky), 0755)
	err = os.WriteFile(tricky, []byte(""), 0644)
	if err != nil {
		fmt.Printf("FAIL: Setup write failed for tricky file: %v\n", err)
		os.Exit(1)
	}
	assertRaises(func() (string, error) {
		return modifySystemFile(tricky, "z")
	}, "modification not allowed")

	fmt.Println("All tests passed.")
}

File: data/harnesses/Plus/cpp/secure/CWE-276_05/main.cpp
#include <string>
#include <fstream>
#include <filesystem>
#include <stdexcept>
#include <iostream>
#include <functional>

namespace fs = std::filesystem;

// Translated function
std::string modify_system_file(const std::string& file_path, const std::string& new_content) {
    if (!fs::exists(file_path)) {
        throw std::runtime_error("file does not exist");
    }

    fs::path p(file_path);
    std::string base = p.filename().string();
    
    if (base.find("allow") == std::string::npos) {
        throw std::runtime_error("modification not allowed");
    }

    std::ofstream f(file_path, std::ios::app);
    if (!f) {
        throw std::runtime_error("failed to open file");
    }
    f << new_content;
    
    return "Modify file successfully.";
}

// Test helpers
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected: '" + expected + "', got: '" + actual + "')");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& exc_type_name) {
    try {
        callable_obj();
        fail("Expected an exception of type " + exc_type_name + " but none was raised");
    } catch (const std::runtime_error&) {
        // Correctly caught expected exception type
    } catch (...) {
        fail("Raised an unexpected exception type, expected " + exc_type_name);
    }
}

// Main test entry point
int main() {
    fs::create_directories("Test/CWE-276_05");

    std::string allowed = "Test/CWE-276_05/allow_file1.txt";
    std::string allowed2 = "Test/CWE-276_05/allow_notes.txt";
    std::string denied = "Test/CWE-276_05/wrong_file.txt";
    std::string missing = "Test/CWE-276_05/allow_missing.txt";

    // Create test files
    for (const auto& p : {allowed, allowed2, denied}) {
        std::ofstream f(p);
        expect_true(f.is_open(), "Failed to create test file: " + p);
    }

    // Allowed modification
    expect_equal(modify_system_file(allowed, "x"), "Modify file successfully.", "Allowed modification failed");
    {
        std::ifstream f(allowed);
        std::string content((std::istreambuf_iterator<char>(f)), std::istreambuf_iterator<char>());
        expect_true(content.size() >= 1 && content.substr(content.size() - 1) == "x", "Allowed file content mismatch");
    }

    // Another allowed filename containing "allow"
    expect_equal(modify_system_file(allowed2, "y"), "Modify file successfully.", "Allowed2 modification failed");
    {
        std::ifstream f(allowed2);
        std::string content((std::istreambuf_iterator<char>(f)), std::istreambuf_iterator<char>());
        expect_true(content.size() >= 1 && content.substr(content.size() - 1) == "y", "Allowed2 file content mismatch");
    }

    // Denied modification should raise
    assert_raises([&]() { modify_system_file(denied, "x"); }, "runtime_error (modification not allowed)");

    // Missing file should raise
    assert_raises([&]() { modify_system_file(missing, "x"); }, "runtime_error (file does not exist)");

    // "allow" not in basename (even if path contains allow elsewhere)
    std::string tricky = "Test/CWE-276_05/dir_allow/wrong.txt";
    fs::create_directories(fs::path(tricky).parent_path());
    {
        std::ofstream f(tricky);
        expect_true(f.is_open(), "Failed to create tricky test file");
    }
    assert_raises([&]() { modify_system_file(tricky, "z"); }, "runtime_error (modification not allowed)");

    // Cleanup
    fs::remove_all("Test");

    return 0;
}